### DISCUSSION TITLE
Invalidate queries still requires hard refresh

### DISCUSSION THREAD

**Discussion by SmitharyZach** [TanStack/query#3922] [Q&A] [answered]
Hello! I am invalidating a query after preforming an action on the database in my app. This action results in new data and so I invalidate the specific query but I still have to do a hard refresh to see the data. This is an issue as the user would have to know to refresh the page to see the new data. I am using next js and have my app wrapped in a the query client. I can post my code below

app.js
```
            <QueryClientProvider client={queryClient} contextSharing={true}>
                                <CartContext.Provider value={cart}>
                                    <Component {...pageProps} />
                                </CartContext.Provider>
                <ReactQueryDevtools initialIsOpen={false} />
            </QueryClientProvider>
```

hook file

```
import { getActiveOrders } from "api/order";
import { useQuery } from "@tanstack/react-query";

const awaitActiveOrders = async (merchantIdentifier: number) => {
    const data = await getActiveOrders(merchantIdentifier);
    return data;
};

export default function useActiveOrders(merchantIdentifier: number) {
    return useQuery(["getActiveOrders", merchantIdentifier], () =>
        awaitActiveOrders(merchantIdentifier)
    );
}
```

and then the invalidate line

`
queryClient.invalidateQueries(["getActiveOrders"]);
`

please let me know if you see anything I am doing wrong!

**Comment by TkDodo** [comment, marked as answer]
it shouldn't be necessary to perform a hard refresh. can you show a codesandbox reproduction of your problem?

**Comment by SmitharyZach** [comment]
@TkDodo thanks for the response! I think I might have something configured incorrectly. Other areas of my app work correctly when the query is invalidated. I am invalidating the query on a different next js page than where the query lives. Could it be there is a new query client created? Does query invalidation have to happen on the same next js page you are viewing the data in?

**Reply by TkDodo** [reply]
see: https://tkdodo.eu/blog/react-query-fa-qs#why-are-updates-not-shown

**Reply by SmitharyZach** [reply]
@TkDodo thanks! I think I ran into the issue. This particular end point is being cached via the header from the back end
`cache store: true, validFor: 300  // 5 min on content`

the headers set are `cache-control: private, max-age=300` 

Is it possible to override the headers when invalidating a query?

**Reply by SmitharyZach** [reply]
I just ended up setting the axois headers to not allow cache, thanks for the help!

**Reply by TkDodo** [reply]
> Is it possible to override the headers when invalidating a query?

no, you cannot override browser caching headers set by the backend with javascript.

**Reply by SmitharyZach** [reply]
@TkDodo I was able to override the cache from the backend with these headers if you ever need them, seems to do the trick:

```    
"Cache-Control": "no-cache",
    Pragma: "no-cache",
    Expires: "0",
```

**Reply by DreamEcho100** [reply]
> @TkDodo I was able to override the cache from the backend with these headers if you ever need them, seems to do the trick:
> 
> ```
> "Cache-Control": "no-cache",
>     Pragma: "no-cache",
>     Expires: "0",
> ```
In my case, it wasn't related to the backend but the browser cache was taking too long so when I tried adding them on the `headers` of the `fetch` it worked
Thanks

**Comment by 0mppula** [comment]
my axios headers are:
`headers: {
	'Cache-Control': 'no-cache',
	Pragma: 'no-cache',
	Expires: '0',
},`

invalidating queries from another component works on initial load. But when hot reloading invalidating queries no longer works i need to refresh browser? any ideas?

EDIT: found out the fix for my problem my queryClient was being instantiated inside of a component 
`const queryClient = new QueryClient();` I moved it outside of it and it worked checkout: https://tkdodo.eu/blog/react-query-fa-qs#2-the-queryclient-is-not-stable
